### PR TITLE
Deadlock Job Cleanup Updateup

### PR DESCRIPTION
**Pull request by pschoffer**
## Description

There is an edge case that can cause the cleanup job of deadlocked jobs ... well to be deadlocked itself :) Please see the comment in the code for more context.

### Plugin Update: Search

Update deadlock cleanup job for the indexing queue to handle correctly multiple jobs trying to reindex the same object.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [ ] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Steps to Test

This is quite hard to test. I think the state can happen due to a race condition in EP code.

The scenario you want is:
 * 2 jobs in the queue - wp_vip_search_index_queue db table
 * 1 needs to be in state 'running'
 * 1 needs to be in state 'scheduled'

So I guess inserting those to DB directly works:
```
INSERT into wp_vip_search_index_queue (`object_id`, `object_type`, `status`, `index_version`, `scheduled_time`) VALUES (1, 'post', 'scheduled', 1, '0000-00-00'), (1, 'post', 'running', 1, '0000-00-00');
```

Then run:
```
wp eval "do_action('vip_search_queue_sweeper');"
```

wp_vip_search_index_queue table should now have only one job in state scheduled and scheduled time now-ish

### CODE REVIEW

File: search/includes/classes/class-queue.php [hunk 1]
```diff
@@ -696,6 +698,41 @@ public function free_deadlocked_jobs() {
 		}
 	}
 
+	/**
+	 * In some cases there could be multiple jobs for the same object stucked in deadlock state.
```
Comment (rinatkhaziev) [inline search/includes/classes/class-queue.php:702]:
```suggestion
	 * In some cases there could be multiple jobs for the same object stuck in a deadlock state.
```

File: search/includes/classes/class-queue.php [hunk 2]
```diff
@@ -696,6 +698,41 @@ public function free_deadlocked_jobs() {
...
+	 * If we would try to update all of them at once we would break the DB contstraint
```
Comment (rinatkhaziev) [inline search/includes/classes/class-queue.php:703]:
```suggestion
	 * If we would try to update all of them at once we would break the DB constraint
```

**Comment by rinatkhaziev** [comment]
Thanks, looks good! Left a couple of minor typo suggestions.

**Review by netsuso — APPROVED** [review]

**Comment by netsuso** [comment]
stacks-r1572

**Pull request merged**